pkg/secrets-store/utils.go

```diff
@@ -0,0 +1,727 @@
+				log.Errorf("failed to set status, err: %v", err)
+				return false, nil
+			}
+			return true, nil
+		}
+		if err := wait.ExponentialBackoff(wait.Backoff{
+			Steps:    5,
+			Duration: 1 * time.Millisecond,
+			Factor:   1.0,
+			Jitter:   0.1,
+		}, setStatusFn); err != nil {
+			log.Error(err, "max retries for setting status reached")
+			return err
+		}
+	}
+	return nil
+}
+
+// removeK8sObjects deletes K8s secrets based on secretProviderClass spec
+// it should also delete pod info from the secretProviderClass object's byPod status field
+func removeK8sObjects(ctx context.Context, targetPath string, podUID string, files []string, secretObjects []interface{}) error {
+	secretProviderClass := ""
+	namespace := ""
+	deleteStatusFn := func() (bool, error) {
+		item, podNS, err := getItemWithPodID(ctx, podUID)
+		if err == nil && len(podNS) > 0 {
+			secretProviderClass = item.GetName()
+			namespace = podNS
+			if err = deleteStatus(ctx, item, podUID); err != nil {
+				return false, nil
+			}
+			return true, nil
+		}
+		return true, nil
+	}
+
+	if err := wait.ExponentialBackoff(wait.Backoff{
+		Steps:    5,
+		Duration: 1 * time.Millisecond,
+		Factor:   1.0,
+		Jitter:   0.1,
+	}, deleteStatusFn); err != nil {
+		log.Error(err, "max retries for deleting status reached")
+		return err
+	}
+
+	if len(namespace) > 0 && len(secretProviderClass) > 0 {
+		deleteSecretFn := func() (bool, error) {
+			item, err := getSecretProviderItemByName(ctx, secretProviderClass)
+			if err != nil {
+				log.Errorf("failed to get secret provider item, err: %v", err)
+				return false, nil
+			}
+			count := getStatusCount(item)
+			// only delete when no more pods are associated with it
+			if count == 0 {
+				///TODO: we assume all files are mounted from a single secretsproviderclass
+				/// a pod could have multiple volumes pointing to diff secretproviderclass objs
+				for _, s := range secretObjects {
+					secretObject, ok := s.(map[string]interface{})
+					if !ok {
+						continue
+					}
+					secretName, exists, err := unstructured.NestedString(secretObject, "secretName")
+					if err != nil {
+						return false, nil
+					}
+					if !exists {
+						continue
+					}
+					err = deleteK8sSecret(ctx, secretName, namespace)
+					if err != nil {
+						return false, nil
+					}
+				}
+			}
+			return true, nil
+		}
+
+		if err := wait.ExponentialBackoff(wait.Backoff{
+			Steps:    5,
+			Duration: 1 * time.Millisecond,
+			Factor:   1.0,
+			Jitter:   0.1,
+		}, deleteSecretFn); err != nil {
+			log.Error(err, "max retries for deleting secret reached")
+			return err
+		}
+	}
+	return nil
+}
+
+// createOrUpdateK8sSecret creates or updates a K8s secret with data from mounted files
+// If a secret with the same name already exists in the namespace of the pod, it's updated.
+func createOrUpdateK8sSecret(ctx context.Context, name string, namespace string, datamap map[string][]byte, secretType corev1.SecretType) error {
+	c, err := getClient()
```

Comment:
Can we add a comment here that we recreate the client to prevent reading from cache?